src/core/data/transform.h

```diff
@@ -30,7 +30,8 @@ struct Transform {
 
 struct StoreTransform : public Transform {
   virtual ~StoreTransform() {}
-  virtual int32_t target_ndim(int32_t source_ndim) const = 0;
+  virtual int32_t target_ndim(int32_t source_ndim) const         = 0;
+  virtual void return_promoted_dims(std::vector<int32_t>&) const = 0;
```

Comment:
also, I feel this method could have a return value, instead of taking an inout argument.

Reply:
I have changed the name to `find_imaginary_dims`.
However I don't think we can have a return value instead of the argument since we need to process `parent` transform before our transform here: https://github.com/ipdemes/legate.core/blob/collective/src/core/data/transform.h#L58-L59

Reply:
that doesn't make sense. if `parent_` is `nullptr`, you're at the bottom of the stack and you just need to forward what's coming out of the `transform_`.